src/libcrun/ebpf.c

```diff
@@ -293,15 +293,90 @@ bpf_program_complete_dev (struct bpf_program *program, libcrun_error_t *err arg_
   return program;
 }
 
+static int
+ebpf_attach_program (int fd, int dirfd, libcrun_error_t *err)
+{
+  const int MAX_ATTEMPTS = 20;
+  int attempt;
+
+  for (attempt = 0;; attempt++)
+    {
+      cleanup_close int replacefd = -1;
+      union bpf_attr attr;
+      uint32_t progs[2];
+      int ret;
+
+      memset (&attr, 0, sizeof (attr));
+      attr.query.target_fd = dirfd;
+      attr.query.attach_type = BPF_CGROUP_DEVICE;
+      attr.query.prog_cnt = sizeof (progs) / sizeof (progs[0]);
+      attr.query.prog_ids = (uint64_t) &progs;
+
+      ret = bpf (BPF_PROG_QUERY, &attr, sizeof (attr));
+      if (UNLIKELY (ret < 0))
+        return crun_make_error (err, errno, "bpf query");
+
+      if (attr.query.prog_cnt > 1)
+        return crun_make_error (err, 0, "invalid device cgroup state, more than one program installed");
+
+      if (attr.query.prog_cnt == 1)
+        {
+#ifdef BPF_F_REPLACE
+          memset (&attr, 0, sizeof (attr));
+          attr.prog_id = progs[0];
+          replacefd = bpf (BPF_PROG_GET_FD_BY_ID, &attr, sizeof (attr));
+          if (UNLIKELY (replacefd < 0))
+            {
+              if (errno == ENOENT && attempt < MAX_ATTEMPTS)
+                {
+                  /* Another update might have raced and updated, try again.  */
+                  continue;
+                }
+              return crun_make_error (err, errno, "cannot open existing eBPF program");
+            }
+#else
+          return crun_make_error (err, 0, "eBPF program already configured");
```

Comment:
If `BPF_F_REPLACE` isn't supported -- which is the case for most kernels -- you should insert the new program then remove the old one. It won't be atomic, but it'll be much better than this.